### ISSUE TITLE
How to CROP detected rectangle to OCR later?

### ISSUE THREAD

**Issue by TeddyPerkins** [ultralytics/yolov5#834]
## ❔Question
I am trying to process a complex (form + table) image. I would like it to **crop** only few content so then they can be OCR-ed.
Is there any arguments I am missing or code itself. 

I saw a [solution on **Stackoverflow** ](https://stackoverflow.com/questions/44544471/how-to-get-the-coordinates-of-the-bounding-box-in-yolo-object-detection)but it was old YOLO.

Any help would be appreciated, Thanks.
<br/>

## Additional context

_YOLO -> **Crop (How to ?)** -> OCR_

**Comment by PuRgE-CoDeE** [comment]
Find the file attested for your solution.
[Crop.txt](https://github.com/ultralytics/yolov5/files/5117517/Crop.txt)

**Comment by NanoCode012** [comment]
If you pass `--save-txt`, it would save the bounding box info to file.

https://github.com/ultralytics/yolov5/blob/5f07782f569324bf32dbc63441128d4935c9254a/detect.py#L108-L109

You can read the file to crop to those positions if I understand you correctly.

**Comment by TeddyPerkins** [comment]
Thank you Mr.Chanvichet, I will try this and let you know.

**Comment by glenn-jocher** [comment]
@TeddyPerkins you can take `img` in detect.py and slice and dice it however you want.

There is object cropping and saving code in the training dataloader which you can use for reference:
https://github.com/ultralytics/yolov5/blob/a8751e50de94b207c1323c3eda493952cced1d72/utils/datasets.py#L404-L421

**Comment by TeddyPerkins** [comment]
oh thats so cool, thank you for pointing this out too.

**Comment by glenn-jocher** [comment]
@TeddyPerkins sure, no problem. The dataloader extraction use-case is obviously to take an existing dataset and extract it's contents into a classification-style dataset, with class-labelled folders etc all automatically.

If you just want to do this for a few images with detect.py however, you might want to simply look at the second-stage classifier code. This code is latent (not used currently), but it's use cases is to pass YOLOv5 detections through a second stage classifier to reduce FPs. As part of this it naturally crops detected boxes to feed to the classifier:
https://github.com/ultralytics/yolov5/blob/8666bc507ede8af73c1cea415c0a04d6c992d338/detect.py#L80-L83

**Comment by TeddyPerkins** [comment]
Whoa, this makes it so much easier pipelining other tasks.

Thank you so much Mr. Glenn!

**Comment by TeddyPerkins** [comment]
@glenn-jocher, but can I crop trapezoid shapes(Keystone effect) to rectangles though?

**Comment by glenn-jocher** [comment]
@TeddyPerkins last time I checked every shape can be enclosed by a rectangle.

**Comment by shiv-acharii** [comment]
https://github.com/ultralytics/yolov5/issues/2608#issuecomment-809094537

U can refer to my reply in the above link.

**Comment by glenn-jocher** [comment]
@TeddyPerkins Prediction box cropping is now available in YOLOv5 via PR https://github.com/ultralytics/yolov5/pull/2827! PyTorch Hub models can use `results.crop()` or detect.py can be called with the `--save-crop` argument. Example usage:
```
python detect.py --save-crop
```
<img width="1121" alt="Screenshot 2021-04-20 at 23 50 51" src="https://user-images.githubusercontent.com/26833433/115468484-70892500-a233-11eb-89e8-717bef0fa45b.png">

**Comment by palash04** [comment]
For those who want to crop the image from the generated labels txt, I have compiled the code from detect.py to crop, given the labels. 

Following is the code snippet. Call **crop_image** with the associated arguments. 

```
def xywhn2xyxy(x, w=640, h=640, padw=0, padh=0):
    # Convert nx4 boxes from [x, y, w, h] normalized to [x1, y1, x2, y2] where xy1=top-left, xy2=bottom-right
    y = np.copy(x)
    y[..., 0] = w * (x[..., 0] - x[..., 2] / 2) + padw  # top left x
    y[..., 1] = h * (x[..., 1] - x[..., 3] / 2) + padh  # top left y
    y[..., 2] = w * (x[..., 0] + x[..., 2] / 2) + padw  # bottom right x
    y[..., 3] = h * (x[..., 1] + x[..., 3] / 2) + padh  # bottom right y
    return y

def xyxy2xywh(x):
    # Convert nx4 boxes from [x1, y1, x2, y2] to [x, y, w, h] where xy1=top-left, xy2=bottom-right
    y = np.copy(x)
    y[..., 0] = (x[..., 0] + x[..., 2]) / 2  # x center
    y[..., 1] = (x[..., 1] + x[..., 3]) / 2  # y center
    y[..., 2] = x[..., 2] - x[..., 0]  # width
    y[..., 3] = x[..., 3] - x[..., 1]  # height
    return y

def xywh2xyxy(x):
    # Convert nx4 boxes from [x, y, w, h] to [x1, y1, x2, y2] where xy1=top-left, xy2=bottom-right
    y = np.copy(x)
    y[..., 0] = x[..., 0] - x[..., 2] / 2  # top left x
    y[..., 1] = x[..., 1] - x[..., 3] / 2  # top left y
    y[..., 2] = x[..., 0] + x[..., 2] / 2  # bottom right x
    y[..., 3] = x[..., 1] + x[..., 3] / 2  # bottom right y
    return y

def clip_boxes(boxes, shape):
    boxes[..., [0, 2]] = boxes[..., [0, 2]].clip(0, shape[1])  # x1, x2
    boxes[..., [1, 3]] = boxes[..., [1, 3]].clip(0, shape[0])  # y1, y2
    return boxes

def crop_image(img, dh, dw, x, y, w, h, gain=1.02, pad=10):
    # img is a numpy type array
    # dh, dw, _ = img.shape
    # class_id, x, y, w, h, conf_score = line
    # where, line is read from labels txt file
    xyxy = xywhn2xyxy(np.array([x,y,w,h]).reshape(1,4), w=dw, h=dh)
    b = xyxy2xywh(xyxy)
    b[:, 2:] = b[:, 2:] * gain + pad
    xyxy = xywh2xyxy(b).astype('int64')
    xyxy = clip_boxes(xyxy, [dh,dw])
    crop = img[int(xyxy[0, 1]):int(xyxy[0, 3]), int(xyxy[0, 0]):int(xyxy[0, 2]), ::(-1)] # keeping -1 as opposed to original code
    
    # save cropped image
    Image.fromarray(crop[..., ::-1]).save('cropped_img.png', quality=95, subsampling=0)  # save RGB
    return crop
```

**Comment by glenn-jocher** [comment]
@palash04 Thanks for sharing the code snippet! This could be a useful reference for anyone looking to crop images from the generated labels. Hope this helps the community!
